src/NHibernate.Test/NHSpecificTest/NH2176/Fixture.cs

```diff
@@ -99,8 +105,42 @@ public ITransaction CreateTransaction(ISessionImplementor session)
 
 		public void EnlistInDistributedTransactionIfNeeded(ISessionImplementor session)
 		{
-			// No enlistment. This disables automatic flushes before ambient transaction
+			// No session enlistment. This disables automatic flushes before ambient transaction
 			// commits. Explicit Flush calls required.
+			// Still make sure the session connection is enlisted, in case it was acquired before 
+			// transaction scope start.
+			// Will not support nested transaction scope. (Will throw, while current NHibernate
+			// just stay in previous scope.)
+			// Will cause an "earlier than required" connection acquisition.
+			// It is required to enlist with null when the scope is ended, otherwise using
+			// the transaction without a new scope will fail by attempting to use it inside
+			// the completed scope.
+			// If an explicit transaction is ongoing, we must not enlist.
+			if (!session.ConnectionManager.Transaction.IsActive)
+			{
+				// Enlist is called terribly frequently, and in some circumstances, it will
+				// not support to be called with the same value. So track what was the previous
+				// call and do not call it again if unneeded.
+				// (And Sql/OleDb/Odbc/Oracle manage/PostgreSql/MySql/Firebird connections support
+				// multiple calls with the same ongoing transaction, but some others may not.)
+				var current = System.Transactions.Transaction.Current;
+				var connection = session.Connection;
+				System.Transactions.Transaction previous;
+				if (!_sessionsTransaction.TryGetValue(connection, out previous) || previous != current)
+				{
+					_sessionsTransaction.AddOrUpdate(connection, current, (s, t) => current);
+					if (current == null &&
+						// This will need an ad-hoc property on Dialect base class instead.
+						(session.Factory.Dialect is SQLiteDialect || session.Factory.Dialect is MsSqlCeDialect))
```

Comment:
Well, it is not a trouble for Sql Server since it supports it in both case. But that is a trouble for SQLite since it will require it if accessed through ODBC (I have just checked, there is an ODBC driver for that) but it will not support it if accessed through its managed driver.